doc/feature_derivation.md

```diff
@@ -0,0 +1,85 @@
+get the background information before going on.
+
+Target: We need to know the below two information for each column after the
+feature type infer routine:
+
+1. How to transform the column data to tensors, including `tf.SparseTensor`.
+2. What type of feature column should adapt to the column and the parameters for the feature column call.
+
+## Simpler COLUMN Clause
+
+When we have a training table contains many columns that should be used
+for training like https://www.kaggle.com/mlg-ulb/creditcardfraud, it's
+not friendly if we must provide all column names in `COLUMN` clause, we'd
+like to just use `COLUMN *` here.
+
+But, for cases when the columns are of quite different data format, like:
+
+```
+data type: (float, float, float, float, csv_string_for_dense_tensor, csv_string_for_sparse_tensor, int)
+column name: (a , b , c , d , e , f, label)
+```
+
+`COLUMN *` and plus the data in the table may not provide enough information we need:
+
+- whether the `csv_string_for_dense_tensor` should going through an embedding layer
+- what the original "dense shape" for the `csv_string_for_sparse_tensor` column to parse to a `SparseTensor`.
+
+So we need SQLFlow users to provide such information in the SQL statements to make the training
+network specific. We support the below SQL statements for the above case:
+
+```
+SELECT * FROM training_table
+TRAIN DNNClassifier
+WITH someattr=somevalue
+COLUMN *,EMBEDDING(e, 128, "sum"),EMBEDDING(SPARSE(f, [1000000]), 512, "sum")
+LABEL label
+INTO my_model_file;
+```
+
+Note that the above SQL have the same meaning to:
+
+```
+SELECT * FROM training_table
+TRAIN DNNClassifier
+WITH someattr=somevalue
+COLUMN a, b, c, d ,EMBEDDING(DENSE(e, [64]), 128, "sum"),EMBEDDING(SPARSE(f, [1000000]), 512, "sum")
+LABEL label
+INTO my_model_file;
+```
+
+Since we can not get the original "dense shape" if the CSV column represents a sparse tensor,
+the shape must be specified in the SQL statement. But if the column represents a "dense tensor"
+we can get the shape by reading some of the values and confirm the shape is the same.
+
+For CSV values, we can also infer the tensor data type by reading some of the training data, whether
+it's int value or float value. Note we can always parse float values to `float32` but not `float64`
+since `float32` seems enough for most cases.
+
+## The Feature Derivation Routine
+
+We need to `SELECT` part of the training data, like 1000 rows and go through the below routine:
+
+1. If the column data type is numeric: int, bigint, float, double, can directly parse to a tensor of shape `[1]`.
+2. If the column data type is string: VARCHAR or TEXT:
+   1. If the string is not one of the supported serialized format (only support CSV currently):
+      1. If all the rows of the column's string data can be parsed to a float or int value,
+         treat it as a tensor of shape `[1]`.
```

Comment:
What if the column data type falls into this category (contents are "1", "2", "3", "4", ...), but a user write `SPARSE(column_name)`. Do we still categorize this column as dense tensor?

Reply:
> I believe the feature derivation logic should start with what feature column a user has written.

You are right.